### ISSUE TITLE
Use two separate dataloaders

### ISSUE THREAD

**Issue by srib** [PyTorchLightning/pytorch-lightning#3336]
### Before asking:   
1. search the issues.   
2. search the docs.    

Here [is](https://github.com/PyTorchLightning/pytorch-lightning/issues/1089) the relevant issue that I have looked up.

#### What is your question?

I have a labeled and an unlabeled dataset that I am using for a semi-supervised segmentation problem. 

Here is what I want to implement in the training loop:

1. Train an epoch on the labeled dataset. (alternatively, a batch)
2. Train an epoch on the unlabeled dataset. (alternatively, a batch)

What is the best way to do it? I modified the code in the issue listed [here](https://github.com/PyTorchLightning/pytorch-lightning/issues/2457).

#### Code

   

Please see below.

#### What have you tried?

```python
    def train_dataloader(self):
        labeled_dataloader = DataLoader(self.labeled_dataset,
                                        batch_size=self.batch_size,
                                        num_workers=4,
                                        shuffle=True, drop_last=True)

        unlabeled_dataloader = DataLoader(self.unlabeled_dataset,
                                          batch_size=self.batch_size,
                                          num_workers=4,
                                          shuffle=True, drop_last=True)
        
        return {'labeled':labeled_dataloader, 'unlabeled': unlabeled_dataloader}
```

#### What's your environment?

 - OS: [e.g. iOS, Linux, Win]
 - Packaging [e.g. pip, conda]
 - Version [e.g. 0.5.2.1]

**Comment by awaelchli** [comment]
Hi, it is currently not possible to return multiple dataloaders for training (that only works for validation). 
A feature for this is in progress here #1959. 

However, in your case, I think it is more elegant to do this:

1. return the right dataloader in each epoch:

```python
def train_dataloader(self):
    if self.current_epoch % 2 == 0:
        labeled_dataloader = ...
        return labeled_dataloader
    else:
        unlabeled_dataloader = ...
        return unlabeled_dataloader
```
2. modify your training_step like this:

```python
def training_step(...):
    if self.current_epoch % 2 == 0:
        # apply loss with labels
    else:
        # apply unsupervised loss
    return ...
```
3. Finally, tell Trainer to call the train_dataloader method every epoch, so it will switch to the new dataset.

```python
trainer = Trainer(..., reload_dataloaders_every_epoch=True)  # False by default
```

**Comment by awaelchli** [comment]
I think this is actually a great question, so I ported it over to the forum to expose it to others who might have the same question.
https://forums.pytorchlightning.ai/t/how-to-switch-dataloaders-between-epochs-in-lightning/137

**Comment by srib** [comment]
Let me try this out. Thank you for the suggestion, @awaelchli!

**Comment by nazim1021** [comment]
Hi @awaelchli   thanks a lot for this solution,  I have a similar problem where i am trying to use unlabeled dataset for semi-supervised classifcation, however i m having trouble on how can we add up the 2 losses and return them as one after each epoch. ? since here the losses are being calculated alternatively

**Comment by awaelchli** [comment]
> how can we add up the 2 losses and return them as one after each epoch

Do you mean iteration? If not, please explain in more detail how your optimization should work. Post your question in the forum and ping me, happy to help.

If you just want to compute two losses from different data and minimize the sum of the two, combine your data into one dataset, for example as described here.
https://pytorch-lightning.readthedocs.io/en/latest/multiple_loaders.html

**Comment by nazim1021** [comment]
Hi @awaelchli  Thanks a lot for prompt response. I have added the question in the forum here -https://forums.pytorchlightning.ai/t/combining-loss-for-multiple-dataloaders/145

**Comment by mariaculman18** [comment]
> Hi, it is currently not possible to return multiple dataloaders for training (that only works for validation).
> A feature for this is in progress here #1959.
> 
> However, in your case, I think it is more elegant to do this:
> 
> 1. return the right dataloader in each epoch:
> 
> ```python
> def train_dataloader(self):
>     if self.current_epoch % 2 == 0:
>         labeled_dataloader = ...
>         return labeled_dataloader
>     else:
>         unlabeled_dataloader = ...
>         return unlabeled_dataloader
> ```
> 
> 1. modify your training_step like this:
> 
> ```python
> def training_step(...):
>     if self.current_epoch % 2 == 0:
>         # apply loss with labels
>     else:
>         # apply unsupervised loss
>     return ...
> ```
> 
> 1. Finally, tell Trainer to call the train_dataloader method every epoch, so it will switch to the new dataset.
> 
> ```python
> trainer = Trainer(..., reload_dataloaders_every_epoch=True)  # False by default
> ```

Hi @awaelchli, I have a question and you might help me. According to documentation, now it is possible to use multiple dataloaders for training (https://pytorch-lightning.readthedocs.io/en/latest/advanced/multiple_loaders.html#multiple-training-dataloaders). But, in training_step, you are required to access each dataloader ->batch_a_b = batch["loaders_a_b"]. So in the end lightning doesn't seem to take care of the batch combination as it claims. Or at least, not in the training part. Any idea on how to feed the dictionary of dataloaders into the training without having to take care of them individually?

**Comment by awaelchli** [comment]
> So in the end lightning doesn't seem to take care of the batch combination as it claims

How do you imagine this "combination" to work? 
If one of the dataloaders returns a batch of shape [16, 32, 32] and the other one returns a batch of shape [16, 3, 32, 32] and a label of shape [16, 1], how do you expect these to get combined? Do you expect a tuple? In the end, you will have to anyway untangle them because you need to feed them to the model(s) somehow.

**Comment by sid-sundrani** [comment]
I was trying to train on multiple dataloaders. Not sure why, but when I fit the trainer again with a new data loader it seems to train on the old dataloader. Number of steps displayed in the former fit is equal to the latter even if its a much smaller dataset

```
trainer.fit(plmodel, train_dataloader, val_dataloader)
trainer.max_epochs += 1
trainer.fit(plmodel, train_dataloader2, val_dataloader)
```

**Comment by awaelchli** [comment]
Yes that's odd, I know. But it's because of these lines of code: 
https://github.com/PyTorchLightning/pytorch-lightning/blob/176df202e4e1e5f5101914929f0f3a3608c41f94/pytorch_lightning/trainer/data_loading.py#L447

Here, if trainer has already attached dataloader(s) from a previous fit, it will not reset them. 

If you set
`trainer.train_dataloader = None`
directly after fit() you will see the problem go away for the next time you call fit().

**Comment by awaelchli** [comment]
I created an issue for this here: #8435
